src/lib/actions/RunAction.ts

```diff
@@ -60,14 +60,14 @@ export class RunAction {
 		await Promise.all(addEnginePromises);
 		const workspace: Workspace = await core.createWorkspace(input.workspace);
 
+		// EngineProgressListeners should start listening right before we call Core's `.selectRules()` method, since
+		// that's when progress events can start being emitted.
+		this.dependencies.progressListeners.forEach(listener => listener.listen(core));
```

Comment:
All listeners start listening at the same time, but they're built to wait silently until they receive their first event, at which point they turn on the spinner. And there's no risk of the spinners colliding, because rule selection must finish before rule execution can begin.